### PR TITLE
Issue vendor CA certificate

### LINKED ISSUE #24: Issue vendor CA certificates
This certificate should be part of the RegisterVendor event and is used by the vendor to issue certificates. In future the certificate will be issued by the Nuts foundation, but for now it's self-signed.

The node administrator then takes the new event file (which has been written to disk) and submits it to the registry on Github.

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes #24

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,6 +13,7 @@ data
 
 # used in test
 tmp
+temp
```
Comment (woutslakhorst) [inline .gitignore]: tmp or temp?

File: docs/_static/nuts-registry.yaml [hunk 2]
```diff
@@ -248,6 +248,10 @@ components:
           example: Medicare Software for People B.V.
         identifier:
           $ref: "#/components/schemas/Identifier"
+        domain:
```
Comment (woutslakhorst) [inline docs/_static/nuts-registry.yaml]: Let's define an enum with at least: healthcare, personal and insurance

(personal being the general term for pgo)

Comment (woutslakhorst) [reply docs/_static/nuts-registry.yaml]: and make it a model

Comment (woutslakhorst) [reply docs/_static/nuts-registry.yaml]: shouldn't it be a mandatory param or default to 'healthcare' (backwards compatible)?

File: docs/pages/configuration/registry.rst [hunk 3]
```diff
@@ -15,16 +15,17 @@ By default it uses the config in the master branch.
...
+registry.mode                            server                                                                                                  server or client, when client it uses the HttpClient
+registry.address                         localhost:1323                                                                                          Interface and port for http server to bind to
+registry.syncMode                        fs                                                                                                      ``fs`` or ``github``
+registry.syncInterval                    30                                                                                                      Interval in minutes to check for new registry data on github
+registry.syncAddress                     https://codeload.github.com/nuts-foundation/nuts-registry-development/tar.gz/master                     The tar.gz to download from github
+registry.vendorCACertificateValidity     365                                                                                                     Validity of the vendor CA certificates in days
```
Comment (woutslakhorst) [inline docs/pages/configuration/registry.rst]: In the future this is determined by the foundation CA. Is it valid to hardcode it until then?

File: pkg/admin.go [hunk 4]
```diff
@@ -1,20 +1,76 @@
...
+		Keys:       []interface{}{jwkAsMap},
 	})
+	err = r.EventSystem.PublishEvent(event)
+	if err != nil {
+		return nil, err
+	}
+	return event, err
+}
+
+func marshalJwk(certASN1 []byte, certType cert.CertificateType) (map[string]interface{}, error) {
```
Comment (woutslakhorst) [inline pkg/admin.go]: Can't the JWK package do this for you given a x509 cert?

File: pkg/cert/x509.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		},
+		ExtraExtensions: extensions,
+	}, nil
+}
+
+// IssueCertificate issues an X.509 certificate to entity 'subject' through Certificate Authority 'ca'. It assumes
+// the CA is under control of the application since it expects the crypto module to directly issue the certificate.
+// Both the subject's and CA's key pair should be available in the crypto module. If subject and CA are equal,
+// it issues a self-signed certificate. Otherwise, the CA's certificate should also be present in the crypto module.
+func IssueCertificate(crypt crypto.Client, csrTemplateFn func() (x509.CertificateRequest, error),
```
Comment (woutslakhorst) [inline pkg/cert/x509.go]: Shouldn't this be in Crypto?

If Crypto has a Vault backend, the actual signing will be done by Vault, right?

File: .gitignore (hunk 1 above)
Comment (reinkrul) [reply .gitignore]: I actually fixed this altogether by getting the test to write in /tmp, so I'll remove these 2

File: docs/pages/configuration/registry.rst (hunk 3 above)
Comment (reinkrul) [reply docs/pages/configuration/registry.rst]: True, I'll hardcode it.

File: pkg/admin.go (hunk 4 above)
Comment (reinkrul) [reply pkg/admin.go]: Just checked, couldn't find it...

File: docs/_static/nuts-registry.yaml (hunk 2 above)
Comment (reinkrul) [reply docs/_static/nuts-registry.yaml]: Fixed

**Comment by reinkrul** [comment]
> Some minor things and 1 real important one: shouldn't be the x509 logic be in crypto?

Well, that depends. Logic to create a CSR and sign it should normally be in the client application which 'owns' the certificate and key pair. However, if there's code there which can be reused by other engines, we probably should move it to nuts-crypto since it also contains x509/crypto helper functions. This might be true for everything in the x509 package (which contains some helper functions around crypto.IssueCertificate() and builders for the CSR), but I'm not convinced (yet) that there will be other engines who'll want to reuse this functionality. So I'd suggest to postpone that decision util then.

File: pkg/cert/x509.go (hunk 5 above)
Comment (reinkrul) [reply pkg/cert/x509.go]: Yes, when you see nuts-registry as Registration Authority, it merely invokes the Certificate Authority (in this case nuts-crypto) to issue a certificate. When certificate signing is done in Vault (accessed through nuts-crypto), Vault will be our CA. If we merely store keys in Vault, I regard nuts-crypto as CA and vault merely as secure storage for keys (a kind of 'software HSM'...).

**Pull request merged**